src/common/src/array/compact_chunk.rs

```diff
@@ -0,0 +1,342 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet};
+
+use crate::array::data_chunk_iter::PkRef;
+use crate::array::{Op, RowRef, StreamChunk};
+use crate::buffer::BitmapBuilder;
+use crate::util::chunk_coalesce::DataChunkBuilder;
+
+/// Compact a chunk by modifying the ops and the visibility of a stream chunk. All UPDATE INSERT and
+/// UPDATE DELETE will be converted to INSERT and DELETE, and dropped according to certain rules
+/// (see `compact_insert` and `compact_delete` for more details).
+pub fn compact_chunk(stream_chunk: StreamChunk) -> StreamChunk {
```

Comment:
I suggest we using a different term from `compact` to avoid confusion with the compaction on the visibility.